src/main/java/org/elasticsearch/bootstrap/Security.java

```diff
@@ -0,0 +1,160 @@
+            
+            // add permissions for all configured paths.
+            Set<Path> paths = new HashSet<>();
+            paths.add(environment.homeFile());
+            paths.add(environment.configFile());
+            paths.add(environment.logsFile());
+            paths.add(environment.pluginsFile());
+            paths.add(environment.workFile());
+            paths.add(environment.workWithClusterFile());
+            for (Path path : environment.dataFiles()) {
+                paths.add(path);
+            }
+            for (Path path : environment.dataWithClusterFiles()) {
+                paths.add(path);
+            }
+            output.write(createPermissions(paths));
+        }
+        return processed;
+    }
+    
+    // package private for testing
+    static byte[] createPermissions(Set<Path> paths) throws IOException {
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        
+        // all policy files are UTF-8:
+        //  https://docs.oracle.com/javase/7/docs/technotes/guides/security/PolicyFiles.html
+        try (Writer writer = new OutputStreamWriter(stream, StandardCharsets.UTF_8)) {
+            writer.write(System.lineSeparator());
+            writer.write("grant {");
+            writer.write(System.lineSeparator());
+            for (Path path : paths) {
+                // add each path twice: once for itself, again for files underneath it
+                addPath(writer, encode(path), "read,readlink,write,delete");
+                addRecursivePath(writer, encode(path), "read,readlink,write,delete");
+            }
+            
+            // on *nix, try to grant read perms to file stores / SSD detection
+            if (!Constants.WINDOWS) {
+                Set<String> stores = new HashSet<>();
+                for (FileStore store : PathUtils.getDefaultFileSystem().getFileStores()) {
+                    try {
+                        String mount = NodeEnvironment.getMountPoint(store);
+                        // mount point for fstat() calls against it
+                        if (mount.startsWith("/")) {
+                            stores.add(mount);
+                        }
+                        // block device: add it for SSD detection
+                        if (store.name().startsWith("/")) {
+                            stores.add(store.name());
+                        }
+                    } catch (Throwable t) {
+                        // these are hacks that are not guaranteed
+                    }
+                }
+                for (String store : stores) {
+                    addPath(writer, encode(store), "read,readlink");
+                }
+                addRecursivePath(writer, "/sys/block", "read,readlink");
+                addRecursivePath(writer, "/sys/devices", "read,readlink");
+                addRecursivePath(writer, "/dev", "read,readlink");
+                addRecursivePath(writer, "/devices", "read,readlink");
+            }
```

Comment:
I want to nuke this entire terrifying 'if' statement and just have the filestore info in Environment (with the other paths) instead of NodeEnvironment. I will open a separate PR for this.